### PR TITLE
Add unionWith

### PR DESCRIPTION
**Pull request by Zelenya**
It's seems like it could be useful. Right now I use a couple of newtypes with different semigroup instances for this

### CODE REVIEW

File: src/Foreign/Object.purs [hunk 1]
```diff
@@ -262,6 +263,12 @@ values = toArrayWithKey (\_ v -> v)
 union :: forall a. Object a -> Object a -> Object a
 union m = mutate (\s -> foldM (\s' k v -> OST.poke k v s') s m)
 
+-- | Compute the union of two maps. If a key exists in both maps,
+-- | its value is computed using the supplied function
```
Comment (kl0tl) [inline src/Foreign/Object.purs]: Could we reuse the documentation from [Data.Map.unionWith](https://github.com/purescript/purescript-ordered-collections/blob/v1.6.1/src/Data/Map/Internal.purs#L617-L618) here?

> Compute the union of two maps, using the specified function to combine values for duplicate keys.

File: src/Foreign/Object.purs [hunk 2]
```diff
@@ -262,6 +263,12 @@ values = toArrayWithKey (\_ v -> v)
...
+unionWith :: forall a. (a -> a -> a) -> Object a -> Object a -> Object a
+unionWith f m1 m2 =
+  mutate (\s1 -> foldM (\s2 k v2 -> OST.poke k (runFn4 _lookup v2 (\v1 -> f v2 v1) k m2) s2) s1 m1) m2
```
Comment (kl0tl) [inline src/Foreign/Object.purs]: Could you rename v2 to v1 (this binding actually denotes values of m1) and v1 to v2 (this binding actually denotes values from m2)?

File: test/Test/Foreign/Object.purs [hunk 3]
```diff
@@ -190,6 +192,24 @@ objectTests = do
...
+        <?> ("k: " <> show k <> ", v1: " <> show v1 <> ", v2: " <> show v2 <> ", key doesn't appear in the union object")
+      Just v ->
+        let expected = unsafePartial fromJust $ lift2 f v1 v2 <|> v1 <|> v2
+        in v == expected
+          <?> ("k: " <> show k <> ", v1: " <> show v1 <> ", v2: " <> show v2 <> ", expected: " <> show expected <> ", actual: " <> show v)
+
+  log "union = unionWith const"
+  quickCheck $ \(TestObject m1 :: TestObject Int) (TestObject m2 :: TestObject Int) ->
+    O.union m1 m2 == O.unionWith const m1 m2
+      <?> (show (O.union m1 m2) <> " != " <> show (O.unionWith const m1 m2))
```
Comment (kl0tl) [inline test/Test/Foreign/Object.purs]: Could we keep the test suite similar to the one of ordered-collections by copying those tests: https://github.com/purescript/purescript-ordered-collections/blob/v1.6.1/test/Test/Data/Map.purs#L205-L224?

**Review by kl0tl — COMMENTED** [review]
This would allow us to implement the Semigroup instance with `append = unionWith (<>)` but this is breaking (see https://github.com/purescript/purescript-foreign-object/pull/19).

**Review by kl0tl — APPROVED** [review]
Thank you!

**Comment by kl0tl** [comment]
We merged https://github.com/purescript/purescript-foreign-object/pull/19 so we could implement the Semigroup instance with `append = unionWith (<>)` in this pull request.

Tests should pass with a rebase once https://github.com/purescript/purescript-foreign-object/pull/22 is merged.

**Review by hdgarrood — APPROVED** [review]

**Pull request merged**